oauth.go

```diff
@@ -0,0 +1,123 @@
+package twilio
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+	"github.com/twilio/twilio-go/client"
+	iam "github.com/twilio/twilio-go/rest/iam/v1"
+)
+
+// TokenAuth handles token-based authentication using OAuth.
+type TokenAuth struct {
+	// token is the cached OAuth token.
+	token string
+	// OAuth is the OAuth client used to fetch new tokens.
+	OAuth client.OAuth
+}
+
+// NewTokenAuth creates a new TokenAuth instance with the provided token and OAuth client.
+func (t *TokenAuth) NewTokenAuth(token string, o client.OAuth) *TokenAuth {
+	return &TokenAuth{token: token, OAuth: o}
+}
+
+// FetchToken retrieves the current token if it is valid, or fetches a new token using the OAuth client.
+func (t *TokenAuth) FetchToken() (string, error) {
+	expired, err := t.Expired()
+	if err != nil {
+		return "", err
+	}
+	if t.token != "" && expired {
+		return t.token, nil
+	}
+
+	token, err := t.OAuth.GetAccessToken(context.Background())
+	if err != nil {
+		return "", err
+	}
+
+	t.token = token
+	return t.token, nil
+}
+
+// Expired checks if the current token is expired.
+func (t *TokenAuth) Expired() (bool, error) {
+	token, _, err := new(jwt.Parser).ParseUnverified(t.token, jwt.MapClaims{})
+	if err != nil {
+		fmt.Printf("Error parsing token: %v\n", err)
+		return true, err
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		fmt.Println("Error asserting token claims")
+		return true, err
+	}
+
+	exp, ok := claims["exp"].(float64)
+	if !ok {
+		fmt.Println("Expiration time (exp) not found in token claims")
+		return true, err
+	}
+
+	expirationTime := time.Unix(int64(exp), 0).UTC()
+	return time.Now().UTC().After(expirationTime), nil
+}
+
+// OAuthCredentials holds the necessary credentials for OAuth authentication.
+type OAuthCredentials struct {
+	// GrantType specifies the type of grant being used for OAuth.
+	GrantType string
+	// ClientId is the identifier for the client application. ClientSecret is the secret key for the client application.
+	ClientId, ClientSecret string
+}
+
+// APIOAuth handles OAuth authentication for the Twilio API.
+type APIOAuth struct {
+	// iamService is the service used to interact with the IAM API.
+	iamService *iam.ApiService
+	// creds holds the necessary credentials for OAuth authentication.
+	creds *OAuthCredentials
+}
+
+// SetIamService sets the iamService field.
+func (a *APIOAuth) SetIamService(service *iam.ApiService) {
```

Comment:
```suggestion
func (a *APIOAuth) SetIAMService(service *iam.ApiService) {
```

Same rule about initialisms. https://google.github.io/styleguide/go/decisions#initialisms